### PR TITLE
Pipeline: use `notify` instead of `polling` for `SharedQueue`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #8869

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by SeaRise** [comment]
wait for https://github.com/pingcap/tiflash/pull/8872

File: dbms/src/Flash/Pipeline/Schedule/Tasks/PipeConditionVariable.h [hunk 1]
```diff
@@ -90,7 +90,10 @@ class PipeConditionVariable
         task->notify();
         task->profile_info.elapsedWaitForNotifyTime();
         assert(TaskScheduler::instance);
-        TaskScheduler::instance->submitToCPUTaskThreadPool(std::move(task));
+        if (unlikely(task->getQueryExecContext().isCancelled()))
+            TaskScheduler::instance->submitToWaitReactor(std::move(task));
+        else
+            TaskScheduler::instance->submitToCPUTaskThreadPool(std::move(task));
```
Comment (SeaRise) [inline dbms/src/Flash/Pipeline/Schedule/Tasks/PipeConditionVariable.h:96]:
```suggestion
        TaskScheduler::instance->submitToCPUTaskThreadPool(std::move(task));
```

File: dbms/src/Operators/Operator.cpp [hunk 2]
```diff
@@ -172,7 +168,8 @@ OperatorStatus SinkOp::prepare()
 OperatorStatus SinkOp::write(Block && block)
 {
     CHECK_IS_CANCELLED
-    profile_info.anchor(block);
+    profile_info.anchor();
+    profile_info.updateInfoFromBlock(block);
```
Comment (SeaRise) [inline dbms/src/Operators/Operator.cpp:172]:
```suggestion
    profile_info.anchor(block);
```

File: dbms/src/Operators/GetResultSinkOp.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <Flash/Executor/ResultQueue.h>
```
Comment (SeaRise) [inline dbms/src/Operators/GetResultSinkOp.cpp:15]:
```suggestion
```

File: dbms/src/Flash/Executor/PipelineExecutorContext.h [hunk 4]
```diff
@@ -26,10 +26,15 @@
 #include <exception>
 #include <mutex>
 
+
```
Comment (SeaRise) [inline dbms/src/Flash/Executor/PipelineExecutorContext.h:29]:
```suggestion
```

File: dbms/src/Flash/Executor/PipelineExecutorContext.cpp [hunk 5]
```diff
@@ -14,8 +14,10 @@
 
 #include <Flash/Executor/PipelineExecutorContext.h>
 #include <Flash/Pipeline/Schedule/TaskScheduler.h>
+#include <Operators/SharedQueue.h>
 
 #include <exception>
+#include <type_traits>
```
Comment (SeaRise) [inline dbms/src/Flash/Executor/PipelineExecutorContext.cpp:20]:
```suggestion
```

File: dbms/src/Operators/Operator.cpp [hunk 6]
```diff
@@ -109,10 +109,7 @@ OperatorStatus SourceOp::read(Block & block)
     assertOperatorStatus(op_status, {OperatorStatus::HAS_OUTPUT});
 #endif
     exec_context.triggerAutoSpill();
-    if (op_status == OperatorStatus::HAS_OUTPUT)
-        profile_info.update(block);
-    else
-        profile_info.update();
+    op_status == OperatorStatus::HAS_OUTPUT ? profile_info.update(block) : profile_info.update();
```
Comment (SeaRise) [inline dbms/src/Operators/Operator.cpp:112]:
```suggestion
    if (op_status == OperatorStatus::HAS_OUTPUT)
        profile_info.update(block);
    else
        profile_info.update();
```

File: dbms/src/Operators/Operator.cpp [hunk 7]
```diff
@@ -131,10 +128,7 @@ OperatorStatus TransformOp::transform(Block & block)
     assertOperatorStatus(op_status, {OperatorStatus::NEED_INPUT, OperatorStatus::HAS_OUTPUT});
 #endif
     exec_context.triggerAutoSpill();
-    if (op_status == OperatorStatus::HAS_OUTPUT)
-        profile_info.update(block);
-    else
-        profile_info.update();
+    op_status == OperatorStatus::HAS_OUTPUT ? profile_info.update(block) : profile_info.update();
```
Comment (SeaRise) [inline dbms/src/Operators/Operator.cpp:131]:
```suggestion
    if (op_status == OperatorStatus::HAS_OUTPUT)
        profile_info.update(block);
    else
        profile_info.update();
```

File: dbms/src/Operators/Operator.cpp [hunk 8]
```diff
@@ -154,10 +148,7 @@ OperatorStatus TransformOp::tryOutput(Block & block)
...
```
Comment (SeaRise) [inline dbms/src/Operators/Operator.cpp:151]:
```suggestion
    if (op_status == OperatorStatus::HAS_OUTPUT)
        profile_info.update(block);
    else
        profile_info.update();
```

File: dbms/src/Common/LooseBoundedMPMCQueue.h [hunk 9]
```diff
@@ -235,12 +306,11 @@ class LooseBoundedMPMCQueue
         queue.pop_back();
         current_auxiliary_memory_usage -= element.memory_usage;
         assert(!queue.empty() || current_auxiliary_memory_usage == 0);
-        writer_head.notifyNext();
         return element.data;
     }
 
     template <typename U>
-    ALWAYS_INLINE void pushFront(U && data)
+    ALWAYS_INLINE bool pushFront(U && data)
```
Comment (gengliqi) [inline dbms/src/Common/LooseBoundedMPMCQueue.h:313]: How about adding comments about the return value?

**Review by gengliqi — APPROVED** [review]

File: dbms/src/Common/LooseBoundedMPMCQueue.h (hunk 9 above)
Comment (SeaRise) [reply dbms/src/Common/LooseBoundedMPMCQueue.h:314]: Has added comment `If returns true, then notify writers afterward; if false, no need to notify writers.`.

File: dbms/src/Common/LooseBoundedMPMCQueue.h [hunk 10]
```diff
@@ -235,12 +306,12 @@ class LooseBoundedMPMCQueue
         queue.pop_back();
         current_auxiliary_memory_usage -= element.memory_usage;
         assert(!queue.empty() || current_auxiliary_memory_usage == 0);
-        writer_head.notifyNext();
```
Comment (windtalker) [inline dbms/src/Common/LooseBoundedMPMCQueue.h:238]: This change will move the `notifyNext` out of the critical area, is it an expected change?

Comment (gengliqi) [reply dbms/src/Common/LooseBoundedMPMCQueue.h:238]: `WaitingNode::notifyNext` is changed to `std::condition_variable::notify_one` which does not need a lock.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**